File: ShaderTypesMemoryLayout.swift
import MetalKit
import simd

print(String(format: "vector_float3: %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "MTLPackedFloat3: %d, %d", MemoryLayout<MTLPackedFloat3>.size, MemoryLayout<MTLPackedFloat3>.stride))
print(String(format: "  x: %d, %d", MemoryLayout<Float>.size, MemoryLayout<Float>.stride))
print(String(format: "  y: %d, %d", MemoryLayout<Float>.size, MemoryLayout<Float>.stride))
print(String(format: "  z: %d, %d", MemoryLayout<Float>.size, MemoryLayout<Float>.stride))
print(String(format: "Camera: %d, %d", MemoryLayout<Camera>.size, MemoryLayout<Camera>.stride))
print(String(format: "  position: %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "  right:    %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "  up:       %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "  forward:  %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "AreaLight: %d, %d", MemoryLayout<AreaLight>.size, MemoryLayout<AreaLight>.stride))
print(String(format: "  position: %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "  forward:  %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "  right:    %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "  up:       %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "  color:    %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "Uniforms: %d, %d", MemoryLayout<Uniforms>.size, MemoryLayout<Uniforms>.stride))
print(String(format: "  width:      %d, %d", MemoryLayout<UInt32>.size, MemoryLayout<UInt32>.stride))
print(String(format: "  heigth:     %d, %d", MemoryLayout<UInt32>.size, MemoryLayout<UInt32>.stride))
print(String(format: "  frameIndex: %d, %d", MemoryLayout<UInt32>.size, MemoryLayout<UInt32>.stride))
print(String(format: "  lightCount: %d, %d", MemoryLayout<UInt32>.size, MemoryLayout<UInt32>.stride))
print(String(format: "  camera: %d, %d", MemoryLayout<Camera>.size, MemoryLayout<Camera>.stride))
print(String(format: "    position: %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "    right:    %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "    up:       %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "    forward:  %d, %d", MemoryLayout<vector_float3>.size, MemoryLayout<vector_float3>.stride))
print(String(format: "Sphere: %d, %d", MemoryLayout<Sphere>.size, MemoryLayout<Sphere>.stride))
print(String(format: "  origin:        %d, %d", MemoryLayout<MTLPackedFloat3>.size, MemoryLayout<MTLPackedFloat3>.stride))
print(String(format: "  radiusSquared: %d, %d", MemoryLayout<Float>.size, MemoryLayout<Float>.stride))
print(String(format: "  color:         %d, %d", MemoryLayout<MTLPackedFloat3>.size, MemoryLayout<MTLPackedFloat3>.stride))
print(String(format: "  radius:        %d, %d", MemoryLayout<Float>.size, MemoryLayout<Float>.stride))
print(String(format: "Triangle: %d, %d", MemoryLayout<Triangle>.size, MemoryLayout<Triangle>.stride))
print(String(format: "  normals: %d, %d",
    MemoryLayout<(vector_float3, vector_float3, vector_float3)>.size,
    MemoryLayout<(vector_float3, vector_float3, vector_float3)>.stride))
print(String(format: "  colors:  %d, %d",
    MemoryLayout<(vector_float3, vector_float3, vector_float3)>.size,
    MemoryLayout<(vector_float3, vector_float3, vector_float3)>.stride))

struct Camera {
    var position: vector_float3
    var right:    vector_float3
    var up:       vector_float3
    var forward:  vector_float3
}

struct AreaLight {
    var position: vector_float3
    var forward:  vector_float3
    var right:    vector_float3
    var up:       vector_float3
    var color:    vector_float3
}

struct Uniforms {
    var width:      UInt32
    var height:     UInt32
    var frameIndex: UInt32
    var lightCount: UInt32
    var camera:     Camera
}

struct Sphere {
    var origin:        MTLPackedFloat3
    var radiusSquared: Float
    var color:         MTLPackedFloat3
    var radius:        Float
}

struct Triangle {
    var normals: (vector_float3, vector_float3, vector_float3)
    var colors:  (vector_float3, vector_float3, vector_float3)
}

File: ShaderTypesMemoryLayout.cpp
/*
 * g++ -I"$(cygpath -w '/cygdrive/c/program files/nvidia gpu computing toolkit/cuda/v11.4/include')" ShaderTypesMemoryLayout.cpp -o ShaderTypesMemoryLayout.exe ; ./ShaderTypesMemoryLayout
 */

#include <stdio.h>
#include <stddef.h>

#include "cuda_runtime.h"
#ifndef __METAL_VERSION__
struct packed_float3 {
#ifdef __cplusplus
    packed_float3() = default;
    packed_float3(float3 v) : x(v.x), y(v.y), z(v.z) {}
#endif
    float x;
    float y;
    float z;
};
#endif

struct Camera {
    float3 position;
    float3 right;
    float3 up;
    float3 forward;
};

struct AreaLight {
    float3 position;
    float3 forward;
    float3 right;
    float3 up;
    float3 color;
};

struct Uniforms {
    unsigned int width;
    unsigned int height;
    unsigned int frameIndex;
    unsigned int lightCount;
    Camera camera;
};

struct Sphere {
    packed_float3 origin;
    float radiusSquared;
    packed_float3 color;
    float radius;
};

struct Triangle {
    float3 normals[3];
    float3 colors[3];
};

int main() {
    printf("float3: %d\n", sizeof(float3));
    printf("packed_float3: %d\n", sizeof(packed_float3));
    printf("  x: %d\n", offsetof(struct packed_float3, x));
    printf("  y: %d\n", offsetof(struct packed_float3, y));
    printf("  z: %d\n", offsetof(struct packed_float3, z));
    printf("Camera: %d\n", sizeof(Camera));
    printf("  position: %d\n", offsetof(struct Camera, position));
    printf("  right:    %d\n", offsetof(struct Camera, right));
    printf("  up:       %d\n", offsetof(struct Camera, up));
    printf("  forward:  %d\n", offsetof(struct Camera, forward));
    printf("AreaLight: %d\n", sizeof(AreaLight));
    printf("  position: %d\n", offsetof(struct AreaLight, position));
    printf("  forward:  %d\n", offsetof(struct AreaLight, forward));
    printf("  right:    %d\n", offsetof(struct AreaLight, right));
    printf("  up:       %d\n", offsetof(struct AreaLight, up));
    printf("  color:    %d\n", offsetof(struct AreaLight, color));
    printf("Uniforms: %d\n", sizeof(Uniforms));
    printf("  width:      %d\n", offsetof(struct Uniforms, width));
    printf("  heigth:     %d\n", offsetof(struct Uniforms, height));
    printf("  frameIndex: %d\n", offsetof(struct Uniforms, frameIndex));
    printf("  lightCount: %d\n", offsetof(struct Uniforms, lightCount));
    int cameraStart = offsetof(struct Uniforms, camera);
    printf("  camera: %d\n", offsetof(struct Uniforms, camera));
    printf("    position: %d\n", cameraStart + offsetof(struct Camera, position));
    printf("    right:    %d\n", cameraStart + offsetof(struct Camera, right));
    printf("    up:       %d\n", cameraStart + offsetof(struct Camera, up));
    printf("    forward:  %d\n", cameraStart + offsetof(struct Camera, forward));
    printf("Sphere: %d\n", sizeof(Sphere));
    printf("  origin:        %d\n", offsetof(struct Sphere, origin));
    printf("  radiusSquared: %d\n", offsetof(struct Sphere, radiusSquared));
    printf("  color:         %d\n", offsetof(struct Sphere, color));
    printf("  radius:        %d\n", offsetof(struct Sphere, radius));
    printf("Triangle: %d\n", sizeof(Triangle));
    printf("  normals: %d\n", offsetof(struct Triangle, normals));
    printf("  colors:  %d\n", offsetof(struct Triangle, colors));
}
